x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlAuthenticateAction.java

```diff
@@ -49,21 +52,30 @@ public TransportSamlAuthenticateAction(
         TokenService tokenService,
         SecurityContext securityContext
     ) {
+        // TODO replace SAME when removing workaround for https://github.com/elastic/elasticsearch/issues/97916
         super(
             SamlAuthenticateAction.NAME,
             transportService,
             actionFilters,
             SamlAuthenticateRequest::new,
-            EsExecutors.DIRECT_EXECUTOR_SERVICE
+            threadPool.executor(ThreadPool.Names.SAME)
         );
         this.threadPool = threadPool;
         this.authenticationService = authenticationService;
         this.tokenService = tokenService;
         this.securityContext = securityContext;
+        this.genericExecutor = threadPool.generic();
     }
 
     @Override
     protected void doExecute(Task task, SamlAuthenticateRequest request, ActionListener<SamlAuthenticateResponse> listener) {
+        // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
+        genericExecutor.execute(ActionRunnable.wrap(listener, l -> doExecuteForked(task, request, l)));
+    }
+
+    private void doExecuteForked(Task task, SamlAuthenticateRequest request, ActionListener<SamlAuthenticateResponse> listener) {
+        assert Transports.assertNotTransportThread("SAML authentication may involve slow IO/HTTP calls");
```

Comment:
Could we have some of these assertions a bit deeper in the SAML codebase please? Clearly we're not on a transport thread here, but we want to be sure we don't accidentally fork back onto one before doing the actually-expensive work.